contrastscan-remote/src/main/java/com/contrastsecurity/scan/ContrastRemoteScanner.kt

```diff
@@ -0,0 +1,193 @@
+            projectId,
+            "scans",
+            scanId
+        )
+        return try {
+            URL(apiUrl.protocol, apiUrl.host, apiUrl.port, path)
+        } catch (e: MalformedURLException) {
+            throw RuntimeException(
+                "Error building clickable Scan URL. Please contact support@contrastsecurity.com for help", e)
+        }
+    }
+
+    private fun waitForResults(scan: Scan, scanOutputPath: Path, timeoutMs: Long) {
+        val scheduler = Executors.newSingleThreadScheduledExecutor()
+        try {
+            val reportsDirectory: Path = scanOutputPath.absolute().normalize().parent
+            try {
+                Files.createDirectories(reportsDirectory)
+            } catch (e: IOException) {
+                throw RuntimeException("Failed to create Contrast Scan reports directory", e)
+            }
+            val await = scan.await(scheduler)
+            val save = await.thenCompose { completed: Scan ->
+                CompletableFuture.runAsync(
+                    {
+                        try {
+                            completed.saveSarif(scanOutputPath)
+                        } catch (e: IOException) {
+                            throw UncheckedIOException(e)
+                        }
+                    },
+                    scheduler
+                )
+            }
+            val output = await.thenAccept { completed: Scan ->
+                val summary: ScanSummary
+                summary = try {
+                    completed.summary()
+                } catch (e: IOException) {
+                    throw UncheckedIOException(e)
+                }
+                writeSummaryToConsole(
+                    summary,
+                    { line: String? -> logger.debug(line) })
+            }
+            CompletableFuture.allOf(
+                save.toCompletableFuture(),
+                output.toCompletableFuture()
+            )[timeoutMs, TimeUnit.MILLISECONDS]
+        } catch (e: ExecutionException) {
+            // try to unwrap the extraneous ExecutionException
+            val cause = e.cause
+            // ExecutionException should always have a cause, but its constructor does not enforce this,
+            // so check if the cause is null
+            val inner = cause ?: e
+            throw RuntimeException("Failed to retrieve Contrast Scan results", inner)
+        } catch (e: InterruptedException) {
+            Thread.currentThread().interrupt()
+            throw RuntimeException("Interrupted while retrieving Contrast Scan results", e)
+        } catch (e: TimeoutException) {
+            val duration = Duration.ofMillis(timeoutMs)
+            val durationString = if (duration.toMinutes() > 0) duration.toMinutes()
+                .toString() + " minutes" else (duration.toMillis() / 1000).toString() + " seconds"
+            throw RuntimeException(
+                "Failed to retrieve Contrast Scan results in $durationString", e
```

Comment:
where are these elaborately created exceptions evaluated?

Reply:
it was originally java code as a maven plugin ported to Kotlin.   Here is a link to the original source https://github.com/Contrast-Security-OSS/contrast-maven-plugin/blob/main/src/main/java/com/contrastsecurity/maven/plugin/ContrastScanMojo.java#L347

I ended up changing this handler to `logger.error()` the detailed string, and then rethrow the original exception.